### DISCUSSION TITLE
Different options for display and input . Can I use a HDMI display and keyboard for input ?

### DISCUSSION THREAD

**Discussion by muziker** [probonopd/MiniDexed#304] [General]
Can I attach a hdmi display and a typing keyboard for display and changing settings?

**Comment by muziker** [comment]
Also I've a gpio touch screen and a usb display that requires some kernel driver compilation. Are there instructions on how to enable these display options? Is it also possible to use a separate usb class compliant sound card for audio?

**Comment by luisfcorreia** [comment]
Hi!

although minidexed is running on a raspberry pi, it does NOT run linux but rather a well crafted C++ program.

So, unless the devices are supported by someone, they won't work.

USB audio is not supported, it requires things which Circle (the underlying library) does nor support.

But, as always, patches are welcome!

**Comment by dolence** [comment]
LVGL seems to be working on circle. That would enable the use of a multitude of displays and beautiful dinamic UI would also be possible, even through HDMI.

**Reply by diyelectromusic** [reply]
The main issue with displays is that MiniDexed uses circle's "character display" modes not the graphical displays, so doing something like this is essentially a complete re-write of the UI.  It's all possible of course... but...

If a touch style interface is of interest, it might be worth looking up MicroDexed-touch which runs on a Teensy: https://codeberg.org/dcoredump/MicroDexed-touch

Kevin

**Reply by dolence** [reply]
I will run some tests later this night. I still have to see how the UI is managed today. But would be nice to have a UI much like the one used by Multidexed. I'm using LVGL on a project from work, but is for embedded STM32.

**Reply by probonopd** [reply]
Hi @dolence, welcome. Are you saying you you are familiar with LVGL? Then we all may be very lucky, since Circle (the framework MiniDexed is using) seems to have support for it:

https://github.com/rsta2/circle/blob/master/addon/lvgl/

There is even a small example:
https://github.com/rsta2/circle/tree/master/addon/lvgl/sample

Personally, I'd find it nice if we could show e.g., the operators on the OLED, [something along these lines](https://github.com/probonopd/MiniDexed/pull/312#issuecomment-1186394819).

**Reply by diyelectromusic** [reply]
> The main issue with displays is that MiniDexed uses circle's "character display" modes not the graphical displays, so doing something like this is essentially a complete re-write of the UI.

**Reply by probonopd** [reply]
If I understand it correctly, LVGL would address the display in graphics mode.
Maybe we can use the current "character display" mode for the whole menu system, but only switch to graphics mode when needed?

**Comment by diyelectromusic** [comment]
> If I understand it correctly, LVGL would address the display in graphics mode.

Yes, but there are two main types of display in circle - text/terminal and graphical.  LVGL and the ST7789 displays are of the graphical variety.  HD44780 and the existing SSD1306 implementation only implement the text-based variety.

There is no code in MiniDexed to make use of the graphical displays in circle - as I say it would require a re-write of the UI.  And the LVGL and ST7789 displays, as I understand it, haven't implemented the character-based API that MiniDexed uses.

One option would be to implement the character API on top of the raw graphics drivers for the ST7789 and LVGL (this is actually what happens for the SSD1306 - it only implements the character interface even though it is a pixel display).  But as far as I know, that hasn't been done yet and that would only give us the existing MiniDexed test-based UI but it would allow the use of more displays.

Kevin

**Reply by probonopd** [reply]
I could imagine that we keep using text mode for 99% of the time and just one some displays (e.g., the larger of the two OLEDs) _occasionally and temporarily_ switch into graphics mode, e.g., when scrolling through the algorithms. Or is that impossible?

Personally I am not interested in more than one display. The point of this device is to be inexpensive and accessible.

**Reply by dolence** [reply]
[LVGL quick test](https://youtu.be/H8eixxSHPPg?si=A5BcrPnKHL4tniE4)

**Reply by diyelectromusic** [reply]
> Or is that impossible?

No, it's all possible, it's just quite a bit of work to make the UI work this way as you need to implement two new things:
* character API support for the graphical displays you want to use which has to really be done in circle.
* graphical support within the user interface of MiniDexed.

And then actually write the graphical user interface of course and some intelligent way to switch between the two - the current UI assumes certain things when it is refreshing at the moment (for example).

Kevin

**Reply by probonopd** [reply]
I was hoping we could simplify things by displaying just icons/symbols __or__ just text at a time, hence not needing a mixture of text and icons/symbols at the same time.

**Reply by diyelectromusic** [reply]
I'm not talking about at the same time - I'm just telling you what is and isn't in the code already.  None of the displays in circle work this way at the moment, so it would need to be added.  Each of the supported displays is /either/ a character /or/ a graphical display. I don't think any of them support both modes.  So unless you are talking about swapping hardware part way through, one or other of the modes needs to be added to the display "drivers" in circle before we can even start thinking about how then to implement the UI in MiniDexed itself.

Everything is possible, just don't underestimate how much work is being implied by the discussion here :)

Kevin

**Comment by dolence** [comment]
We can't select display type by passing an argument? LVGL or character type.
LVGL can create flex designs that adapt to the screen resolution. By default it uses small buffers that update only the needed part of the screen.

**Reply by diyelectromusic** [reply]
It is all possible, I agree, but as I say:

>  doing something like this is essentially a complete re-write of the UI

Kevin

**Comment by dolence** [comment]
I've managed to get the correct resolution. Touch panel still not working. This is the display I will use on my 3D printer. I will try to find a more affordable one. Later I will make a test of the áudio output on the display using MiniDexed and try to get it running alongside.

[Correct resolution](https://youtu.be/qXCMJ39oHBE?si=poSGW-kMP9CKN7ia)

**Comment by dolence** [comment]
I have started working on a touch UI mockup, but I'm kind of lost on how it should really be. Suggestions?

![Performances](https://github.com/probonopd/MiniDexed/assets/5223003/1483a529-9bc1-484c-8b78-415ba0e537d2)

**Reply by probonopd** [reply]
Well, personally I'd me much more interested in something usable with screens we are using in this project, like this 128x64 pixel monochrome OLED display that can be had for ~3 USD:

![image](https://github.com/probonopd/MiniDexed/assets/2480569/da885f1c-8f35-461a-9ba2-837f1d5b40d2)

**Reply by dolence** [reply]
I'm not aiming to change anything on the actual working screen. One could use i2c screens, hdmi (possibly with audio) or both at same time. There are some cheap hdmi screens. By cheap I mean around the price of a Raspberry Pi 3. If someone wanna go cheaper, just stick to the SSD1306. What you think?

**Reply by dolence** [reply]
Later, LVGL support could be added to SSD1306/SSD1309.
[Take a look at this video!](https://www.youtube.com/watch?v=a7K2PldYtTE)

**Reply by probonopd** [reply]
> If someone wanna go cheaper, just stick to the SSD1306. 

That'd be me. 
I do see value in seeing the graphical diagrams for the algorithms on those tiny screens.

There are already plenty of solutions with big color screens, like [MicroDexed](https://codeberg.org/positionhigh/MicroDexed-touch) - no need to reinvent that.
